### ISSUE TITLE
React Native: Invariant Violation with util-crypto

### ISSUE THREAD

**Issue by Tbaut** [polkadot-js/common#921]
Hey there, here I am with a RN question :)

When upgrading `@polkadot/util-crypto` from v5.9.2 to v6.0.2
I run into the following: 

```
Invariant Violation: expected hoisted manifest for "@polkadot/reactnative-identicon#@polkadot/util-crypto#@polkadot/wasm-crypto"
    at invariant (/home/thib/.nvm/versions/node/v14.15.4/lib/node_modules/yarn/lib/cli.js:2314:15)
    at _loop2 (/home/thib/.nvm/versions/node/v14.15.4/lib/node_modules/yarn/lib/cli.js:94526:9)
    at PackageHoister.init (/home/thib/.nvm/versions/node/v14.15.4/lib/node_modules/yarn/lib/cli.js:94596:19)
    at PackageLinker.getFlatHoistedTree (/home/thib/.nvm/versions/node/v14.15.4/lib/node_modules/yarn/lib/cli.js:48744:20)
    at PackageLinker.<anonymous> (/home/thib/.nvm/versions/node/v14.15.4/lib/node_modules/yarn/lib/cli.js:48755:27)
    at Generator.next (<anonymous>)
    at step (/home/thib/.nvm/versions/node/v14.15.4/lib/node_modules/yarn/lib/cli.js:310:30)
    at /home/thib/.nvm/versions/node/v14.15.4/lib/node_modules/yarn/lib/cli.js:328:14
    at new Promise (<anonymous>)
    at new F (/home/thib/.nvm/versions/node/v14.15.4/lib/node_modules/yarn/lib/cli.js:5301:28)
```
All my other @polkadot package are running the latest version (non beta).
Any hint?

**Comment by jacogr** [comment]
According to this 

- https://github.com/staticfuse/create-gatsby-theme-publisher/issues/4
- https://github.com/yarnpkg/yarn/issues/7244

deleting and re-creating a new yarn.lock solves the issue.

**Comment by jacogr** [comment]
Until Monday when the next version of the API will depend on these versions (or probably this weekend), I would suggest adding the util* packages to yarn resolutions as well - just to ensure all resolve to a single/latest version. 

e.g. https://github.com/polkadot-js/apps/blob/master/package.json#L23 - the UI is a bit behind there, but it is only because I have not pushed the updated packages, since there is no functional differences

**Comment by Tbaut** [comment]
I had tried this but it didn't help. I installed `@polkadot/reactnative-identicon` v0.70.2-3 and then tried again with `@polkadot/util-crypto`. Although I was able to install it, it fails when launching the app with:

```
error: Error: Unable to resolve module ./data.cjs from /home/thib/github/stylo-app/stylo/node_modules/@polkadot/wasm-crypto-asmjs/data.js: 

None of these files exist:
  * node_modules/@polkadot/wasm-crypto-asmjs/data.cjs(.native|.android.js|.native.js|.js|.android.json|.native.json|.json|.android.ts|.native.ts|.ts|.android.tsx|.native.tsx|.tsx)
  * node_modules/@polkadot/wasm-crypto-asmjs/data.cjs/index(.native|.android.js|.native.js|.js|.android.json|.native.json|.json|.android.ts|.native.ts|.ts|.android.tsx|.native.tsx|.tsx)
  4 | // Generated as part of the build, do not edit
  5 |
> 6 | export { asmJsInit } from './data.cjs';
    |                            ^
  7 |
    at ModuleResolver.resolveDependency (/home/thib/github/stylo-app/stylo/node_modules/metro/src/node-haste/DependencyGraph/ModuleResolution.js:168:15)
    at DependencyGraph.resolveDependency (/home/thib/github/stylo-app/stylo/node_modules/metro/src/node-haste/DependencyGraph.js:353:43)
    at Object.resolve (/home/thib/github/stylo-app/stylo/node_modules/metro/src/lib/transformHelpers.js:271:42)
    at resolve (/home/thib/github/stylo-app/stylo/node_modules/metro/src/DeltaBundler/traverseDependencies.js:571:33)
    at /home/thib/github/stylo-app/stylo/node_modules/metro/src/DeltaBundler/traverseDependencies.js:587:26
    at Array.reduce (<anonymous>)
    at resolveDependencies (/home/thib/github/stylo-app/stylo/node_modules/metro/src/DeltaBundler/traverseDependencies.js:586:33)
    at /home/thib/github/stylo-app/stylo/node_modules/metro/src/DeltaBundler/traverseDependencies.js:275:33
    at Generator.next (<anonymous>)
    at asyncGeneratorStep (/home/thib/github/stylo-app/stylo/node_modules/metro/src/DeltaBundler/traverseDependencies.js:87:24)

```

**Comment by Tbaut** [comment]
> Until Monday when the next version of the API will depend on these versions (or probably this weekend), I would suggest adding the util* packages to yarn resolutions as well - just to ensure all resolve to a single/latest version.

Got it, I thought they were all in sync. I have absolutely no rush and can remain on 5.9, there are some warnings of having different versions in parallel but this is no big trouble.

**Comment by jacogr** [comment]
Weird, it is there -

```bash
$ ls -al node_modules/@polkadot/wasm-crypto-asmjs/
total 6544
drwxr-xr-x  12 jaco  staff      384 Mar  4 22:24 .
drwxr-xr-x  28 jaco  staff      896 Mar  2 12:30 ..
-rw-r--r--   1 jaco  staff  3312169 Mar  4 22:24 data.cjs
-rw-r--r--   1 jaco  staff       38 Mar  4 22:24 data.d.ts
-rw-r--r--   1 jaco  staff      202 Mar  4 22:24 data.js
-rw-r--r--   1 jaco  staff      275 Mar  4 22:24 empty.cjs
-rw-r--r--   1 jaco  staff       38 Mar  4 22:24 empty.d.ts
-rw-r--r--   1 jaco  staff      143 Mar  4 22:24 empty.js
-rw-r--r--   1 jaco  staff     1303 Mar  4 22:24 package.json
-rw-r--r--   1 jaco  staff      388 Mar  4 22:24 packageInfo.cjs
-rw-r--r--   1 jaco  staff       78 Mar  4 22:24 packageInfo.d.ts
-rw-r--r--   1 jaco  staff      250 Mar  4 22:24 packageInfo.js
```

**Comment by jacogr** [comment]
So the above may be that in fixing one issue that I've introduced another.... i.e. it doesn't see the `.cjs` extension as something it can resolve.

**Comment by Tbaut** [comment]
I def. have it too...
```
ls -al node_modules/@polkadot/wasm-crypto-asmjs/       
total 3280
drwxr-xr-x  2 thib thib    4096 Mar  5 22:05 .
drwxr-xr-x 23 thib thib    4096 Mar  5 22:02 ..
-rw-r--r--  1 thib thib 3312169 Mar  5 20:19 data.cjs
-rw-r--r--  1 thib thib      38 Mar  5 20:19 data.d.ts
-rw-r--r--  1 thib thib     202 Mar  5 20:19 data.js
-rw-r--r--  1 thib thib     275 Mar  5 20:19 empty.cjs
-rw-r--r--  1 thib thib      38 Mar  5 20:19 empty.d.ts
-rw-r--r--  1 thib thib     143 Mar  5 20:19 empty.js
-rw-r--r--  1 thib thib     388 Mar  5 20:19 packageInfo.cjs
-rw-r--r--  1 thib thib      78 Mar  5 20:19 packageInfo.d.ts
-rw-r--r--  1 thib thib     250 Mar  5 20:19 packageInfo.js
```

**Comment by jacogr** [comment]
See this comment - https://github.com/thysultan/stylis.js/issues/233#issuecomment-663635896

**Comment by jacogr** [comment]
So there are 2 places where `something.js` imports from `something.cjs`, one in wasm and 1 in util - there is a hack I can apply that makes it work, will see what i can do to remove the above. (The whole idea behind these major bumps was to _not_ have these extension issues)

**Comment by Tbaut** [comment]
> Adding to metro config -
> resolver: {
    sourceExts: ['js', 'json', 'ts', 'tsx', 'cjs'],
  },

Thanks a lot, this did the trick, now having those work well together
```
    "@polkadot/reactnative-identicon": "^0.70.2-3",
    "@polkadot/util": "^6.0.2",
    "@polkadot/util-crypto": "^6.0.2",
```

The bundler is sending warning about mixing versions, but that's what you mention, with @polkadot/metadata and type using 3.9 of wasm crypto, but that's what you mentioned before, they are included in api afaik

**Comment by jacogr** [comment]
Ok, have the 2 imports I can think of PR-ed. So that should not be needed afterwards.

**Comment by jacogr** [comment]
When you do have a gap, would appreciate -

- the removal of the cjs extension in the lookup
- upgrade to 6.0.3 - I'm hoping I covered all the bases there and it now "just works"

**Comment by Tbaut** [comment]
Just tried with `util` and `until-crypto` on 6.0.3 and sadly it failed when I removed the resolver in `metro.config.js` (copy/pasting still because given the hour, I might miss something)
```
error: Error: Unable to resolve module ./data.cjs from /home/thib/github/stylo-app/stylo/node_modules/@polkadot/ui-shared/node_modules/@polkadot/wasm-crypto-asmjs/data.js: 

None of these files exist:
  * node_modules/@polkadot/ui-shared/node_modules/@polkadot/wasm-crypto-asmjs/data.cjs(.native|.android.js|.native.js|.js|.android.json|.native.json|.json|.android.ts|.native.ts|.ts|.android.tsx|.native.tsx|.tsx)
  * node_modules/@polkadot/ui-shared/node_modules/@polkadot/wasm-crypto-asmjs/data.cjs/index(.native|.android.js|.native.js|.js|.android.json|.native.json|.json|.android.ts|.native.ts|.ts|.android.tsx|.native.tsx|.tsx)
  4 | // Generated as part of the build, do not edit
  5 |
> 6 | export { asmJsInit } from './data.cjs';
    |                            ^
  7 |
    at ModuleResolver.resolveDependency (/home/thib/github/stylo-app/stylo/node_modules/metro/src/node-haste/DependencyGraph/ModuleResolution.js:168:15)
    at DependencyGraph.resolveDependency (/home/thib/github/stylo-app/stylo/node_modules/metro/src/node-haste/DependencyGraph.js:353:43)
    at Object.resolve (/home/thib/github/stylo-app/stylo/node_modules/metro/src/lib/transformHelpers.js:271:42)
    at resolve (/home/thib/github/stylo-app/stylo/node_modules/metro/src/DeltaBundler/traverseDependencies.js:571:33)
    at /home/thib/github/stylo-app/stylo/node_modules/metro/src/DeltaBundler/traverseDependencies.js:587:26
    at Array.reduce (<anonymous>)
    at resolveDependencies (/home/thib/github/stylo-app/stylo/node_modules/metro/src/DeltaBundler/traverseDependencies.js:586:33)
    at /home/thib/github/stylo-app/stylo/node_modules/metro/src/DeltaBundler/traverseDependencies.js:275:33
    at Generator.next (<anonymous>)
    at asyncGeneratorStep (/home/thib/github/stylo-app/stylo/node_modules/metro/src/DeltaBundler/traverseDependencies.js:87:24)

```

**Comment by jacogr** [comment]
Yes, that is messy deps, if you look at the path it has an own version. So bumping UI-* to 071 (the released version), will solve that one.

TL;DR Same as the one above, just popping out in a different place since this part doesn't have bumped versions. (A yarn.lock rewrite... again, should also adjust, but the bump of the UI is probably simpler)

PS: The same will pop up in API, until released. The good news is that the first usage seems indeed solved, so just the bumps to the latest packages should clean things up. This was really valuable, rather earlier than later.

**Comment by Tbaut** [comment]
Thanks for the followup, indeed with with following and a fresh `yarn.lock` it works well, without the custom resolver:
>  "@polkadot/api": "^3.11.1",
    "@polkadot/reactnative-identicon": "^0.71.1",
    "@polkadot/util": "^6.0.3",
    "@polkadot/util-crypto": "^6.0.3
